### DISCUSSION TITLE
Best Practices for Managing Tenant Data in a Multi-Tenant Application

### DISCUSSION THREAD

**Discussion by JordanAtSatori** [oqtane/oqtane.framework#4890] [Q&A]
Hi all,

I'm working on a multi-tenant application and am currently handling tenant data from within a master tenant. Below is a simplified example of my current approach using dependency injection to retrieve tenant-specific configurations.

I'm wondering if there's a standard or best-practice way to manage tenant data in scenarios like this. Specifically, is there a more efficient or secure approach to handle database connection strings?

```
@inject ITenantService TenantService
@inject IConfigManager ConfigManager

@foreach (var tenant in tenants)
{
    string dbConnectionString = ConfigManager.GetConnectionString(tenant.DBConnectionString);
    Console.WriteLine($"Name: {tenant.Name}");
    Console.WriteLine($"DBConnectionName: {tenant.DBConnectionString}");
    Console.WriteLine($"DBConnectionString: {dbConnectionString}");
    Console.WriteLine($"DBType: {tenant.DBType}");
    Console.WriteLine($"TenantId: {tenant.TenantId}");
    Console.WriteLine($"Version: {tenant.Version}");
}

@code {
    List<Tenant> tenants = new();
    protected override async Task OnInitializedAsync()
    {
        tenants = await TenantService.GetTenantsAsync();
    }
}
```

**Comment by sbwalker** [comment]
@JordanAtSatori can you please explain what you mean by "handling tenant data"... are you referring specifically to accessing the data in the Tenant table in the database and the associated connection strings in appsettings.json, or are you referring to accessing data from any table within an Oqtane database? In general, developers should not need to access the tenant or connection string information ie. the idea behind multi-tenancy is that these details are abstracted. 

Some comments on the code you included above:
- ConfigManager is a Server API which means that the service will only be available if the UI component is running on the server ie. as a Static component or using Interactive / Server configuration
- if you want to follow a client/server approach then I would suggest you look at Oqtane.Client\Modules\Admin\Sql\Index.razor as it loads the list of connection strings using the SystemInfo API which works in both client and server configurations

**Comment by JordanAtSatori** [comment]
Instead of logging into each individual tenant to manage tenant-specific options, we want to centralize the management of these configurations within a master tenant. For example, each tenant has settings stored in their tenant-specific database that define which API endpoints they can access. Managing these settings from a centralized master tenant allows us to make changes more efficiently, without the overhead of logging into each tenant separately.

Am I understanding correctly that the SystemInfo API you mentioned would be suitable for accessing and managing these settings across tenant databases from the master tenant?

**Comment by sbwalker** [comment]
@JordanAtSatori apologies for the delay in responding. The Setting API in Oqtane has the ability to store configuration for various entities. It sounds like you want to be able to store configuration data for each Tenant. You could use the Setting API for this purpose:

Task<Dictionary<string, string>> GetTenantSettingsAsync()
Task UpdateTenantSettingsAsync(Dictionary<string, string> tenantSettings)

You might find this explanation of data scoping useful as well: https://www.oqtane.org/blog/!/72/data-scoping
